### ISSUE TITLE
Weird recursive type in Babbage's rules

### ISSUE THREAD

**Issue by KtorZ** [IntersectMBO/cardano-ledger#2843]
Hi, I just wanted to point out that I found this particular type _awkward_ and potentially dangerous / error-prone:

https://github.com/input-output-hk/cardano-ledger/blob/95d9f20887495a3a6f8d61a675016b888bf5d186/eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs#L105-L120

I get that this is embedding predicate failures from Alonzo for re-use, but the hierarchy seems quite wrong / recursive. We can see that the predicate failures for both `"UTXO"` and `"UTXOW"` rules point to the same type:

https://github.com/input-output-hk/cardano-ledger/blob/95d9f20887495a3a6f8d61a675016b888bf5d186/eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs#L456

https://github.com/input-output-hk/cardano-ledger/blob/95d9f20887495a3a6f8d61a675016b888bf5d186/eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxow.hs#L271

---

This makes the `BabbageUtxoPred` a recursive type. Indeed, the second branch contains a `UtxowPredicateFail era` coming from the Alonzo's rules:

https://github.com/input-output-hk/cardano-ledger/blob/95d9f20887495a3a6f8d61a675016b888bf5d186/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxow.hs#L106-L135

This one also wraps era failures from the Shelley era as:

```hs
WrappedShelleyEraFailure !(UtxowPredicateFailure era)
```

Now, looking into [Shelley's UtxowPredicateFailure](https://github.com/input-output-hk/cardano-ledger/blob/1db68a3ec0a2dcb5751004beb22b906162474f23/eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Utxow.hs#L156-L179), we can see that it is also a bigger type itself embedding the following:

https://github.com/input-output-hk/cardano-ledger/blob/1db68a3ec0a2dcb5751004beb22b906162474f23/eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Utxow.hs#L166

Which is where it gets interesting, because we do have the following:

```hs
PredicateFailure (Core.EraRule "UTXO" (BabbageEra crypto)) ~ BabbageUtxoPred (BabbageEra crypto)
```

This makes the `BabbageUtxoPred` a recursive data-structure containing itself. I found it quite weird / dangerous because it is far from obvious at first read and, I am flagging this because that's the first time I see this kind of construction in the ledger rules. I guess it all stems from the fact that the type used as predicate failures for the `UTXO` and `UTXOW` rules are the same. It wouldn't be the case if those types were different. So my intuition is that the first branch of `BabbageUtxoPred` should not be there, but instead, we should have:

```hs
type PredicateFailure (BabbageUTXO era) = UtxoPredicateFailure era 
```

in the era rules machinery for Babbage (where this `UtxoPredicateFailure` corresponds to the Alonzo one).

cc @JaredCorduan

**Comment by JaredCorduan** [comment]
Thanks for bringing this up @KtorZ (and for the effort it took to make this issue)!

I had a look with @TimSheard just now, we agree that it is a bit weird, but are unclear how it is dangerous. Tim's intuition as that we should replace

https://github.com/input-output-hk/cardano-ledger/blob/1db68a3ec0a2dcb5751004beb22b906162474f23/eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Utxow.hs#L166

with

```
  | UtxoFailure (UtxoPredicateFailure era)
```

I started along this path, there are a lot of cascading constraint changes that need to be made. But we'll let you know how it goes. Without a clear reason as to why it is dangerous, we'll need to get any changes in ASAP if it is to make it into the hardfork changes.

**Comment by KtorZ** [comment]
I find it error-prone, or _dangerous_ in the sense that, functions that would digest values of that type may actually loop forever. It is theoretically possible to create a `FromAlonzoUtxowFail (FromAlonzoUtxowFail (FromAlonzoUtxowFail (FromAlonzoUtxowFail (...` and so forth since the type does not really enforce any termination. I notice this while writing the following function:

```hs
encodeUtxoFailure
    :: Crypto crypto
    => Rules.Shelley.PredicateFailure (Ledger.EraRule "UTXO" (BabbageEra crypto))
    -> Json
encodeUtxoFailure = \case
    Ledger.Babbage.FromAlonzoUtxowFail e ->
        Alonzo.encodeUtxowPredicateFail encodeUtxoFailure e
``` 

This makes me quite uncomfortable because, depending on how the `BabbageUtxoPred` was constructed, this function may never terminates. 

From a rapid look at how the predicate failures are currently constructed, I think it is _fine_; but I am not 100% sure given that I only looked quickly at it and the machinery is far from trivial here.

**Comment by JaredCorduan** [comment]
The strategy I described above does not work, by the way. Shelley-MA relies on being able to swap out the `UTXO` predicate failure (while keeping the same `UTXOW` predicate failure).

I also tried the strategy described in the main issue, namely setting `type PredicateFailure (BabbageUTXO era) = UtxoPredicateFailure era ` in the babbage UTXO STS instance. I time-boxed the effort to a couple hours, didn't finish, but maybe it would work. It's certainly a lot to get one's head around all the predicate failures. I would love to clean it all up (maybe one top-level, era-specific failure type?), but this will take some effort to come up with a strategy for not breaking the node-to-client serialization.

Another option is to employ code copy, but that is disappointing as well.

There are no loops in our current design, but I get that it's not friendly code.
